File: BackApi/BackApi.Web/Controllers/ChangePasswordController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System;
using System.ComponentModel.DataAnnotations;
using Users.App;

namespace BackApi.Web.Controllers
{
    [Route("/")]
    public class ChangePasswordController(ChangePasswordService changePasswordService) : ControllerBase
    {
        [Authorize]
        [HttpPut("changepassword")]
        public async Task<ActionResult> ChangePassword([FromBody] ChangePasswordRequestDto request)
        {
            if (!ModelState.IsValid)
            {
                return BadRequest(ModelState);
            }

            var result = await changePasswordService.ChangePasswordAsync(new ChangePasswordRequest
            {
                UserId = request.UserId,
                OldPassword = request.OldPassword,
                NewPassword = request.NewPassword,
                LockEnabled = true
            });

            if (!result.Success)
            {
                return result.Error switch
                {
                    "UserNotFound" => NotFound(new { error = result.Error, error_description = result.ErrorDescription }),
                    "InvalidOldPassword" => Unauthorized(new { error = result.Error, error_description = result.ErrorDescription }),
                    "AccountLocked" => Forbid(),
                    "InvalidAccountType" => BadRequest(new { error = result.Error, error_description = result.ErrorDescription }),
                    "InvalidPasswordFormat" => BadRequest(new { error = result.Error, error_description = result.ErrorDescription }),
                    _ => BadRequest(new { error = result.Error ?? "ChangePasswordFailed", error_description = result.ErrorDescription })
                };
            }

            return Ok(new { message = "Password changed successfully." });
        }
    }

    internal static class Validation
    {
        public const string PasswordPattern = @"^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[^A-Za-z0-9]).{8,}$";
        public const string PasswordFormatError = "Password must be at least 8 characters long and contain at least one uppercase letter, one lowercase letter, one digit, and one special character.";
    }

    public record ChangePasswordRequestDto(
        [Required(ErrorMessage = "User ID is required.")]
        Guid UserId,
        [Required(AllowEmptyStrings = false, ErrorMessage = "Old password is required")]
        string OldPassword,
        [Required(AllowEmptyStrings = false, ErrorMessage = "New password is required")]
        [RegularExpression(Validation.PasswordPattern, ErrorMessage = Validation.PasswordFormatError)]
        string NewPassword
    );
}

File: NodeBackApi/src/controllers/changePasswordController.ts
import { Router, type Request, type Response } from "express";

import { requireJwt } from "../middleware/jwtAuth";
import { ChangePasswordService } from "../services/ChangePasswordService";
import { passwordFormatError, passwordPattern } from "../utils/validation";

type ChangePasswordControllerDeps = {
  changePasswordService: ChangePasswordService;
};

type ChangePasswordBody = {
  oldPassword?: string;
  newPassword?: string;
  OldPassword?: string;
  NewPassword?: string;
};

const getField = (body: ChangePasswordBody, ...keys: Array<keyof ChangePasswordBody>): string => {
  for (const key of keys) {
    const value = body[key];
    if (typeof value === "string") {
      return value;
    }
  }
  return "";
};

export const buildChangePasswordRouter = ({
  changePasswordService,
}: ChangePasswordControllerDeps): Router => {
  const router = Router();

  router.put(
    "/changepassword",
    requireJwt,
    async (req: Request, res: Response): Promise<void> => {
      const body = (req.body ?? {}) as ChangePasswordBody;
      const oldPassword = getField(body, "oldPassword", "OldPassword");
      const newPassword = getField(body, "newPassword", "NewPassword");

      const errors: Record<string, string[]> = {};
      if (!oldPassword) {
        errors.OldPassword = ["Old password is required"];
      }
      if (!newPassword) {
        errors.NewPassword = ["New password is required"];
      } else if (!passwordPattern.test(newPassword)) {
        errors.NewPassword = [passwordFormatError];
      }

      if (Object.keys(errors).length > 0) {
        res.status(400).json(errors);
        return;
      }

      const userId = typeof req.auth?.sub === "string" ? req.auth.sub : "";
      if (!userId) {
        res.sendStatus(401);
        return;
      }

      const result = await changePasswordService.changePasswordAsync({
        userId,
        oldPassword,
        newPassword,
        lockEnabled: true,
      });

      if (!result.success) {
        switch (result.error) {
          case "UserNotFound":
            res.status(404).json({
              error: result.error,
              error_description: result.errorDescription,
            });
            return;
          case "InvalidOldPassword":
            res.status(401).json({
              error: result.error,
              error_description: result.errorDescription,
            });
            return;
          case "AccountLocked":
            res.sendStatus(403);
            return;
          case "InvalidAccountType":
          case "InvalidPasswordFormat":
            res.status(400).json({
              error: result.error,
              error_description: result.errorDescription,
            });
            return;
          default:
            res.status(400).json({
              error: result.error ?? "ChangePasswordFailed",
              error_description: result.errorDescription,
            });
            return;
        }
      }

      res.status(200).json({ message: "Password changed successfully." });
    },
  );

  return router;
};
